src/softmax.jl

```diff
@@ -53,38 +71,39 @@ julia> Dense(4 => 7, softmax)(x)
 ERROR: `softmax(x)` called with a number, but it expects an array. 
 ```
 """
-softmax(x::AbstractArray{T}; dims = 1) where {T} = softmax!(similar(x, float(T)), x; dims)
+softmax(x::AbstractArray{T}; kw...) where {T} = softmax!(similar(x, float(T)), x; kw...)
 
-softmax!(x::AbstractArray; dims = 1) = softmax!(x, x; dims)
+softmax!(x::AbstractArray; kw...) = softmax!(x, x; kw...)
 
-function softmax!(out::AbstractArray{T}, x::AbstractArray; dims = 1) where {T}
+function softmax!(out::AbstractArray{T}, x::AbstractArray; dims = 1, mask = true) where {T}
+    mask isa AbstractArray{Bool} || mask === true || error("invalid mask, only boolean arrays are accepted")
     max_ = fast_maximum(x; dims)
     if all(isfinite, max_)
-        @fastmath out .= exp.(x .- max_)
+        @fastmath out .= mask .* exp.(x .- max_)
```

Comment:
ideally the mask should be applied before computing the maximum, otherwise there is the risk that all unmasked exponentials are very small. Do you think that applying it like in this line is still numerically stable?

Reply:
It fails if there is `Inf`. I haven't thought hard beyond that.

Can try some variants doing it before.